### ISSUE TITLE
TypeScript Compiler does not understand resulting types when using &&

### ISSUE THREAD

**Issue by Ristaaf** [Microsoft/TypeScript#12174]
**TypeScript Version:**  2.0.8

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
let s:string=("" && false);
```

**Expected behavior:**
Since ("" && false) is "" in JavaScript I should be able to assign the resulting value to a string, but TypeScript thinks that this will become a boolean.
**Actual behavior:** 
Compiler error claiming that a boolean cannot be assigned to a string

Disclaimer: I think that the behaviour of the && operator in JavaScript is horrible, however it is what it is and TypeScript should honour it.

**Comment by RyanCavanaugh** [comment]
Agree that's not the desired behavior, but did you find this writing real code?

**Comment by ahejlsberg** [comment]
Your example compiles with no error using the nightly build. The issue was fixed with our improved support for literal types in #10676.

**Comment by Ristaaf** [comment]
RyanCavanaough: This was found when turning on strictNullChecks on an existing codebase, not the example that I showed of course, that was just a simplification to be clear. 

We had complex logical expressions with && that relied on returning a string. They could have returned other types too, but we did not handle that, so to understand what was going on and to fix the code, we started to read up on the && operator and what it returns in different circumstances, then we tried those circumstances out in TypeScript and found the error reported here.
